### PR TITLE
KEYCLOAK-15955 Implement a pluggable Upgrade strategy

### PR DESCRIPTION
**Pull request by vmuzikar**

### CODE REVIEW

**Review by slaskawi — CHANGES_REQUESTED** [review]
LGTM, @andymunro May I kindly ask you to review and merge this one?

**Comment by vmuzikar** [comment]
In the latest force-push I've fixed the failing tests.

File: server_installation/topics/operator/upgrading.adoc [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+=== Upgrade strategy
+
+You can configure how the operator performs {project_name} upgrades. You can choose from the following upgrade strategies.
+
+* `recreate`: This is the default strategy. The operator removes all {project_name} replicas, optionally creates a backup
+  and then creates the replicas based on a newer {project_name} image. This strategy is suitable for major upgrades as
+  a single {project_name} version is accessing the underlying database. The downside is {project_name} needs to be shut
+  down during the upgrade.
+* `rolling`: The operator removes one replica at a time and creates it again based on a newer {project_name} image. This
+  ensures a zero-downtime upgrade but is more suitable for minor version upgrades that don't require database migration
```
Comment (andymunro) [inline server_installation/topics/operator/upgrading.adoc]: Avoid contractions.  Change "don't" to "do not"

**Review by andymunro — CHANGES_REQUESTED** [review]
Looks good, Sebastian. Just one comment.

**Comment by vmuzikar** [comment]
@andymunro Thanks for the review! Fixed.

**Comment by vmuzikar** [comment]
The test failure is unrelated to this.

**Comment by vmuzikar** [comment]
Rebased to resolve conflicts.
@andymunro Could we please merge this?

**Review by stianst — APPROVED** [review]

**Comment by stianst** [comment]
Approving based on previous reviews from Andy and Sebastian

**Pull request merged**